### PR TITLE
Autodetect what language a query is for

### PR DESCRIPTION
**Pull request by shati-patel**
Adds a `findLanguage` function that uses `codeql resolve queries --format bylanguage` to detect a query's language. This might not work for some queries, so I've added a manual selection as a fallback.

This is used in two places:
- When running a query on multiple databases (we now only show the databases that correspond to the query language)
- When running a remote query (we can now autodetect a language when it isn't specified in the `.repositories` config. This config file is going to be replaced by a different UI soon.)

Question: Does this need tests? I wasn't sure where/how to go about that!

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
   - Only the "multiple DBs" part is user-facing for now
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/docs-content-codeql` has been cc'd in all issues for UI or other user-facing changes made by this pull request. 
   - No docs update required (beyond pending update from https://github.com/github/codeql/pull/6322)

### CODE REVIEW

File: extensions/ql-vscode/src/cli.ts
Comment (shati-patel) [inline extensions/ql-vscode/src/cli.ts:490]: Findings so far:
1. `findLanguage` works fine for all standard queries I've tried in the `<lang>/ql/src` folder
2. It fails for other queries, e.g. 
    - quick queries
    - example snippets in `<lang>/ql/examples/snippets`
    - [custom examples](https://github.com/github/vscode-codeql-starter/tree/main/codeql-custom-queries-cpp) in the starter workspace

For those failures, `codeql` seems unable to resolve the dependency in the qlpack 🤔 

<details><summary>Expand for log</summary>



```
Finding query language using CodeQL CLI: resolve queries -v --log-to-stderr --format bylanguage [redacted]\vscode-codeql-starter\ql\cpp\ql\examples\snippets\addressof.ql...
Sending shutdown request
Sent shutdown request
[2021-07-26 15:11:39] Calling plumbing command: codeql resolve library-path --qlconfig-file=[redacted]\AppData\Local\Programs\Microsoft VS Code\qlconfig.yml --query=[redacted]\vscode-codeql-starter\ql\cpp\ql\examples\snippets\addressof.ql --find-extractors --full-library-path=none --dbscheme=none --format=json
[2021-07-26 15:11:39] [DETAILS] resolve library-path> Resolving query at normalized path [redacted]\vscode-codeql-starter\ql\cpp\ql\examples\snippets\addressof.ql.
[2021-07-26 15:11:39] [DETAILS] resolve library-path> Found enclosing pack 'codeql-cpp-examples' at [redacted]\vscode-codeql-starter\ql\cpp\ql\examples.
[2021-07-26 15:11:39] [DETAILS] resolve library-path> Adding compilation cache at [redacted]\vscode-codeql-starter\ql\cpp\ql\examples\.cache.
[2021-07-26 15:11:39] [DETAILS] resolve library-path> Resolving library dependencies from [redacted]\vscode-codeql-starter\ql\cpp\ql\examples\qlpack.yml.
[2021-07-26 15:11:39] [ERROR] resolve library-path> ERROR: Dependency not found: codeql-cpp ([redacted]\vscode-codeql-starter\ql\cpp\ql\examples\qlpack.yml:1,1-1)
A fatal error occurred: Could not resolve library path for [redacted]\vscode-codeql-starter\ql\cpp\ql\examples
[2021-07-26 15:11:39] Exception caught at top level: Could not resolve library path for [redacted]\vscode-codeql-starter\ql\cpp\ql\examples
                      com.semmle.cli2.resolve.ResolveLibraryPathCommand.executeJSON(ResolveLibraryPathCommand.java:232)
                      com.semmle.cli2.resolve.ResolveLibraryPathCommand.executeJSON(ResolveLibraryPathCommand.java:45)
                      com.semmle.cli2.picocli.SimpleJsonSubcommand.executeInsistingOnJSON(SimpleJsonSubcommand.java:104)
                      com.semmle.cli2.picocli.PlumbingRunner.call(PlumbingRunner.java:54)
                      com.semmle.cli2.picocli.SubcommandCommon.callPlumbingInProcess(SubcommandCommon.java:123)
                      com.semmle.cli2.resolve.ResolveQueriesCommand.splitByLanguage(ResolveQueriesCommand.java:867)
                      com.semmle.cli2.resolve.ResolveQueriesCommand.printNonJsonResult(ResolveQueriesCommand.java:841)
                      com.semmle.cli2.resolve.ResolveQueriesCommand.printNonJsonResult(ResolveQueriesCommand.java:63)
                      com.semmle.cli2.picocli.SimpleJsonSubcommand.executeSubcommand(SimpleJsonSubcommand.java:84)
                      com.semmle.cli2.picocli.SubcommandCommon.executeWithParent(SubcommandCommon.java:432)
                      com.semmle.cli2.execute.CliServerCommand.lambda$executeSubcommand$0(CliServerCommand.java:67)
                      com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:205)
                      com.semmle.cli2.execute.CliServerCommand.executeSubcommand(CliServerCommand.java:67)
                      com.semmle.cli2.picocli.SubcommandCommon.call(SubcommandCommon.java:448)
                      com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:205)
                      com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:214)
                      com.semmle.cli2.CodeQL.main(CodeQL.java:98)
```
</details>

Comment (shati-patel) [reply extensions/ql-vscode/src/cli.ts:490]: Woops, looks like I just needed to specify some `--additional-packs` 😄

File: extensions/ql-vscode/src/cli.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/cli.ts:502]: I would slightly lean towards adding a new type like we have for other CLI commands (e.g. `TestCompleted`) to capture the output of `codeql resolve queries --format bylanguage`) and then using that result type rather than doing this.

Comment (edoardopirovano) [reply extensions/ql-vscode/src/cli.ts:502]: To elaborate a bit on why I think that's better: Capturing the whole result of the comment rather than just doing some ad-hoc parsing of it here means that if in future we want to call that command for some other reason that uses another part of the output, it will be much easier to do.

Comment (shati-patel) [reply extensions/ql-vscode/src/cli.ts:502]: That makes sense, thanks! 🎉 I'll have a go at capturing the result as a type 😃

File: extensions/ql-vscode/src/cli.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/cli.ts:50]: Yeah, it's currently just a placeholder and is only ever an empty object. If `unknown` below is enough for the JSON deserializer to be happy then that seems as good a placeholder as any for the moment.

Comment (shati-patel) [reply extensions/ql-vscode/src/cli.ts:50]: Great, thank you! (And thanks for the review in general ⚡ ) 

I'll convert this back to draft while I tidy up the rest 😀

File: extensions/ql-vscode/src/cli.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/cli.ts:496]: nit: I don't think _analyzing_ is the right word. But I'm not really sure what the best word is. This suggestion is possibly, slightly better.

```suggestion
   * Finds the language that a query targets.
```

**Review by aeisenberg — COMMENTED** [review]
This looks right to me!

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/run-remote-query.ts:43]: I notice that in this case we continue to the next line and return an empty string. It might be better to change the type of this function to `string | undefined` and then places where this function are called will be forced to handle the `undefined` case explicitly. In particular, I think you'll want to:
1. When filtering databases, simply skip the filtering and show all databases.
2. When running queries remotely, not make the API call without a language.

**Review by edoardopirovano — COMMENTED** [review]
Mostly looks good apart from a minor comment inline. I'll hold off on approving it, though, since we paired on this for a bit so it should probably have another set of eyes.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/run-remote-query.ts:31]: Before you return this value, I suggest saving it in a `const` variable and logging a message. That way when looking at the extension log, we can immediately tell whether we followed this code path and successfully detected a language.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/run-remote-query.ts:41]: Could this ever be `undefined` rather than the empty string? If yes, `if(!language)` will handle both cases.

File: extensions/ql-vscode/src/cli.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/cli.ts:504]: Since you asked about tests:
- `extensions/ql-vscode/src/vscode-tests/cli-integration/run-cli.test.ts` is a good place to add integration tests for the `resolveQueryByLanguage` and `resolveLanguages` functions.
- You can use the existing tests as an example.
- `resolveLanguages` is probably easier to write a test for because there's no query as input. 
- `resolveQueryByLanguage` will require a test query: there are some examples in the `data` folders, but as far as I can tell none of them depend on an actual language. So you may need to add one, or save this test for later.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/run-remote-query.ts:23]: Since you asked about tests: I agree that having tests for this function would be handy, and give you confidence in its behaviour. However we don't yet have a test module for `run-remote-query` due to the experimental nature of the entire feature, so adding a test requires some upfront work. (In future it will be easier :) )

I'll record some pointers here about how you could create those tests, but I'm not expecting you to do it in this PR. We can look at test coverage together after this PR.
- For integration tests (which use a real query and actually run the underlying commands), copy `extensions/ql-vscode/src/vscode-tests/cli-integration/queries.test.ts` as a template. I think it would make sense to have:
  - one test function that runs `findLanguage` on a query with a real language dependency and identifies the language
  - one test function that runs it on a dummy query and returns `undefined`
  - any other cases you can think of
- For unit tests (which don't run the actual commands, but simulate the underlying behaviour), use `extensions/ql-vscode/src/vscode-tests/no-workspace/query-results.test.ts` as a template. Those tests mock the behaviour of the CLI server.

**Review by adityasharad — COMMENTED** [review]
Looks good! I think the code is mostly ready to go, with one outstanding suggestion from Edoardo about return types, and one from me about logging.

I've added some suggestions about tests, but I think those are fine to add later, since it will require some upfront work to set up tests for this part of the code.

Have we been adding changelog entries for this feature?

**Comment by shati-patel** [comment]
Thank you for the reviews ✨  I've addressed the main comments in https://github.com/github/vscode-codeql/pull/915/commits/83c68ecbf89415c462c83ac9fd4da8efbe2a2f39 and added a minimal changelog entry that covers the user-facing part of this.

I'll look at adding tests in a separate PR!

File: extensions/ql-vscode/src/extension.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/extension.ts:583]: Note that if the user has no databases and doesn't pick a language when prompted, we try to display this prompt with nothing in it. There should probably be an extra check to display an error in this case and return, similarly to what happens with the filtered list above.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/run-remote-query.ts:74]: I would add a brief comment stating that we don't need an error message here because `findLanguage` already displays one.

**Review by edoardopirovano — COMMENTED** [review]
A couple more minor comments, but otherwise looks good!

**Review by adityasharad — APPROVED** [review]

File: extensions/ql-vscode/src/extension.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/extension.ts:587]: Woops, good point! Thanks 😄

**Review by edoardopirovano — APPROVED** [review]
Thanks for addressing all my comments, this LGTM 👍🏼

**Pull request merged**
